drivers/gpu/drm/virtio/virtgpu_plane.c

```diff
@@ -239,6 +239,41 @@ static int virtio_gpu_check_plane_scaling(struct drm_plane_state *state,
 	return 0;
 }
 
+static unsigned int adjusted_rate(const struct drm_rect *src,
+				 const struct drm_rect *dst,
+				 unsigned int rate)
+{
+	unsigned int src_w, src_h, dst_w, dst_h;
+
+	src_w = drm_rect_width(src) >> 16;
+	src_h = drm_rect_height(src) >> 16;
+	dst_w = drm_rect_width(dst);
+	dst_h = drm_rect_height(dst);
+
+	/* Downscaling limits the maximum pixel rate */
+	dst_w = min(src_w, dst_w);
+	dst_h = min(src_h, dst_h);
+
+	return DIV_ROUND_UP_ULL(mul_u32_u32(rate, src_w * src_h),
+				dst_w * dst_h);
+}
+
+static int virtio_gpu_plane_cdclk_check(struct drm_plane_state *plane_state,
+		                         struct drm_crtc_state *crtc_state)
+{
+	unsigned int plane_cdclk;
+	unsigned int pixel_rate = adjusted_rate(&plane_state->uapi.src,
+			&plane_state->uapi.dst, crtc_state->mode.crtc_clock);
+
+	plane_cdclk = DIV_ROUND_UP(pixel_rate, 2);
+
+	printk("%s pixelrate :%lu plane cdclk:%lu \r\n", __func__, pixel_rate, plane_cdclk);
+
+	if(plane_cdclk >= 652800) /* igpu max cdclk */
```

Comment:
the max cdclk depends on specific generation of GPU. 652800 as I checked is Gen10's (Cannon Lake) max. For RPL which is Gen12, it is 810000+? could you double confirm?